BeatSaberHTTPStatus/HTTPServer.cs

```diff
@@ -59,24 +63,38 @@ public class HTTPServer {
 
 	public class StatusBroadcastBehavior : WebSocketBehavior {
 		private StatusManager statusManager;
+		private Task readyToWrite = Task.CompletedTask;
+		private readonly CancellationTokenSource connectionClosed = new CancellationTokenSource();
 
 		public void SetStatusManager(StatusManager statusManager) {
 			this.statusManager = statusManager;
-
 			statusManager.statusChange += OnStatusChange;
 		}
 
-		protected override void OnOpen() {
-			JSONObject eventJSON = new JSONObject();
+		/// <summary>Queue data to send on the websocket in-order. This method is thread-safe.</summary>
+		private void QueuedSend(string data) {
+			var promise = new TaskCompletionSource<object>();
+			var oldReadyToWrite = Interlocked.Exchange(ref readyToWrite, promise.Task);
+			oldReadyToWrite.ContinueWith(t => {
+				SendAsync(data, b => {
```

Comment:
Considering this is already off-thread, is there any reason to use `SendAsync` and schedule yet another task?

Reply:
If there's no in-progress messages being sent at the time, the callback passed to `ContinueWith` will be executed within the main thread when it calls QueuedSend because of the `TaskContinuationOptions.ExecuteSynchronously` flag. (If there is a currently sending message, then the `ContinueWith` callback will later run on the websocket-sharp thread when `promise.SetResult(null)` is called.) Using the synchronous flag means the send can start immediately instead of only happening once a thread in the default thread pool gets around to it. No threads besides the game main thread and whatever websocket-sharp-managed threads are involved, and none of the threads have blocking IO happen on them.

If the `TaskContinuationOptions.ExecuteSynchronously` flag was not present as was the case in an earlier version of this PR, then it's a fair question. The code could instead look like this:

```csharp
		private void QueuedSend(string data) {
			var promise = new TaskCompletionSource<object>();
			var oldReadyToWrite = Interlocked.Exchange(ref readyToWrite, promise.Task);
			oldReadyToWrite.ContinueWith(t => {
				Send(data);
				promise.SetResult(null);
			}, connectionClosed.Token);
		}
```

This would still fix the issue of the game main thread being blocked, but it would have the downsides that the send wouldn't start until a thread from the thread pool was free to execute the callback and picked it up, and doing blocking IO on the default thread pool would tie up a thread for a period of time, possibly increasing the time it takes other thread pool tasks to start and maybe causing more threads to be instantiated. Using async IO would minimize our impact on the threadpool. (And then using the ExecuteSynchronously flag completely removes our impact and dependence on the threadpool.)

Reply:
I see, thanks for explaining!